File: string/leetcode_67.java
package string;

public class leetcode_67 {
    public static void main(String[] args) {
        leetcode_67 lc = new leetcode_67();
        System.out.println(lc.addBinary("11", "1"));
    }

    public String addBinary(String a, String b) {
        StringBuilder sb = new StringBuilder();
        int i = a.length() - 1, j = b.length() - 1, sum = 0;
        while (i >= 0 || j >= 0) {
            sum = sum / 2;
            if (i >= 0) sum += a.charAt(i--) - '0';
            if (j >= 0) sum += b.charAt(j--) - '0';
            sb.append(sum % 2);
        }
        if (sum / 2 != 0) sb.append(1);
        return sb.reverse().toString();
    }
}

File: bit_manipulation/LeetCode_67.kt
package bit_manipulation

class LeetCode_67 {

    fun addBinary(a: String, b: String): String {
        var i = a.lastIndex
        var j = b.lastIndex

        val stringBuilder = StringBuilder()
        var carry = false

        while (i >= 0 || j >= 0) {
            var sum = 0

            if (carry) {
                sum += 1
            }

            if (i >= 0) {
                sum += a[i--].digitToInt()
            }

            if (j >= 0) {
                sum += b[j--].digitToInt()
            }

            carry = sum > 1

            stringBuilder.append(sum % 2)
        }

        if (carry) {
            stringBuilder.append(1)
        }

        return stringBuilder.reverse().toString()
    }
}

fun main() {
    // Input
    val a = "11"
    val b = "1"
    // Output
    LeetCode_67().addBinary(a, b).let {
        println(it)
        require(it == "100")
    }
}
